APE17.rst

```diff
@@ -0,0 +1,482 @@
+tests directly with pytest, e.g.::
+
+    pytest packagename
+
+so using tox is optional.
+
+We note that these changes have no impact on the availability of the
+``package.test()`` function which is unrelated to astropy-helpers and relies
+instead on the astropy core package to provide a test runner.
+
+Building documentation
+~~~~~~~~~~~~~~~~~~~~~~
+
+Astropy-helpers provides a ``python setup.py build_docs`` command that
+behind-the-scenes built the package then added it to ``sys.path``, then ran the
+documentation build. Having the package be importable is needed as documentation
+often includes API sections that are dynamically created based on the package.
+
+As for testing, we can accomplish the same process more cleanly instead using
+`tox`_ - a minimal tox configuration might look like::
+
+    [testenv:build_docs]
+    deps =
+      sphinx-astropy
+    commands =
+      sphinx-build docs docs/_build/html -W -b html
+
+and this would be run using::
+
+    tox -e build_docs
+
+Developers not wishing to use tox could also accomplish the same by doing::
+
+    pip install -e .
+    cd docs
+    make html
+
+These changes will have no impact on ReadTheDocs as that service never made use
+of the ``build_docs`` command, instead invoking sphinx directly with
+``sphinx-build``.
+
+Version helpers
+~~~~~~~~~~~~~~~
+
+Currently, astropy-helpers includes version helpers that take the version
+defined in ``setup.cfg`` or ``setup.py`` and add a developer string when in a
+checked out version of a git repository. The developer string consists of
+``version.devN`` where N is the number of commits since a release.
+
+The Python Packaging Authority (PyPA) now provide a setuptools extension called
+`setuptools_scm`_ which can entirely replace the version helpers in
+astropy-helpers. The way this package works is that versions are no longer
+specified in ``setup.cfg`` or ``setup.py`` - instead the versions are taken from
+tags. Developer version strings produced in this way are much more sophisticated
+and can indicate for example if the working copy is clean or has local changes,
+and whether it is a stable tagged version or a developer version. We note that
+the default way of setting up setuptools_scm results in the version string not
+being updated automatically in ‘editable’ installs of a package (i.e. ``pip
+install -e .``), however `a workaround
+<https://github.com/pypa/setuptools_scm/issues/273>`_ is available.
+
+Switching to `setuptools_scm`_ requires minimal configuration, which is well
+described in its documentation and we therefore do not repeat here. However, one
+subtlety is that since it relies on tags to determine versions, for packages
+such as the core astropy package which have all their tags on branches (at least
+in recent years), we will need to add a ‘developer’ tag on ``master`` straight
+after branching, e.g. after creating a ``v4.0.x`` branch we should tag the next
+commit on ``master`` as ``v4.1.dev``.
```

Comment:
It seems like this should be noted somewhere, though, or it will be forgotten in the future. Maybe part of release instructions?

Reply:
It also seems a bit problematic because it means *affiliated* packages would have to start doing this.  And many affiliated packages do not have standardized release procedures/instructions...

Reply:
No I don't think affiliated packages would need to do this. This issue only occurs for packages that have a workflow where they tag releases in branches, but in my experience most affiliated packages don't do release branches.

Reply:
I think we should leave this in the APE - the APE is *not* documentation but just a technical document for the project, and I think it's important to note this here. But I've added a sentence that says

*In any case, the documentation about how to release the core package as well astropy-affiliated packages will need to be updated to reflect the use of setuptools_scm.*

Reply:
We are including instructions about how to use setuptools_scm with release branches in an upcoming packaging guide.